File: algorithms-sort/Python/heap-sort.py
def HeapSort(index, size_of_vector, vetor):
    max_index = index
    if (2 * index + 1) <= size_of_vector:
        index_right = (2 * index + 1)
        index_left = 2 * index

        if vetor[index_right] >= vetor[index_left] and vetor[index_right] > vetor[index]:
            max_index = (2 * index + 1)
        elif vetor[index_left] > vetor[index_right] and vetor[index_left] > vetor[index]:
            max_index = 2 * index
    elif 2 * index <= size_of_vector:
        index_left = 2 * index
        if vetor[index_left] > vetor[index]:
            max_index = 2 * index

    if max_index != index:
        auxiliar = vetor[index]
        vetor[index] = vetor[max_index]
        vetor[max_index] = auxiliar
        HeapSort(max_index, size_of_vector, vetor)

def transformaHeap(size_of_vector, vetor):
    index = round(size_of_vector/2)
    while index >= 0:
        HeapSort(index, size_of_vector, vetor)
        index -= 1

def ordena(size_of_vector, vetor):
    index = size_of_vector
    while index >= 1:
        vetor[0], vetor[index] = vetor[index], vetor[0]
        last_position = index - 1
        HeapSort(0, last_position, vetor)
        index -= 1

vetor = [10, 3, 4, 2, 5, 6, 8, 7, 9, 1, 11, 14, 13, 102, 0, 44, 56, 33, 12, 103, 99]

print(f'Vetor antes da ordenação:\n{vetor}')

transformaHeap(len(vetor) - 1, vetor)
ordena(len(vetor) - 1, vetor)

print(f'Vetor após a ordenação:\n{vetor}')    

File: algorithms-sort/C++/heap-sort.cpp
#include <iostream>
#include <cmath>

using namespace std;

int vetor[10] = {8,10,4,6,3,2,1,7,9,5};

void HeapSort(int index, int size_of_vector) {
    int index_right,
    index_left,
    maior_index,
    auxiliar;
    
    maior_index = index;

    if(2 * index + 1 <= size_of_vector) {
        index_right = 2 * index + 1;
        index_left = 2 * index;
        if(vetor[index_right] > vetor[index_left] && vetor[index_right] > vetor[index])
        maior_index = 2 * index + 1;
        else if (vetor[index_left] > vetor[index_right] && vetor[index_left] > vetor[index])
        maior_index = 2 * index;
    }
    else if(2 * index <= size_of_vector) {
        index_left = 2 * index;
        if(vetor[index_left] > vetor[index])
        maior_index = 2 * index;
    }

    if(maior_index != index) {
        auxiliar = vetor[index];
        vetor[index] = vetor[maior_index];
        vetor[maior_index] = auxiliar;
        HeapSort(maior_index, size_of_vector);
    }
}

void TransformaHeap(int size_of_vector) {
    int index;
    int half = round(size_of_vector/2);
    for(index =  half; index >= 0; index--) {
        HeapSort(index, size_of_vector);
    }
}

void ordena(int size_of_vector) {
    int index,
    auxiliar,
    last_position;
    for(index = size_of_vector; index >= 1; index--) {
        auxiliar = vetor[0];
        vetor[0] = vetor[index];
        vetor[index] = auxiliar;
        last_position = index - 1;
        HeapSort(0, last_position);
    }
}

void showVetor(int size_of_vector, int vetor[]) {
    int index;
    for(index = 0; index <= size_of_vector; index++) {
        cout << vetor[index] << " ";
    }
}

int main() {

    int size_of_vector = 9;

    cout << "Vetor antes da ordenação:\n";
    showVetor(size_of_vector, vetor);

    TransformaHeap(size_of_vector);
    ordena(size_of_vector);

    cout << "\nVetor após a ordenação:\n";
    showVetor(size_of_vector, vetor);

    return 0;
}
